src/hardware/signal_generator.rs

```diff
@@ -174,32 +174,26 @@ impl core::iter::Iterator for SignalGenerator {
 
     /// Get the next value in the generator sequence.
     fn next(&mut self) -> Option<i16> {
-        self.phase_accumulator = self.phase_accumulator.wrapping_add(
-            if (self.phase_accumulator as i32).is_negative() {
-                self.config.frequency_tuning_word[0]
-            } else {
-                self.config.frequency_tuning_word[1]
-            },
-        );
-
-        let phase = self.phase_accumulator as i32;
+        let sign = self.phase_accumulator.is_negative();
+        self.phase_accumulator = self
+            .phase_accumulator
+            .wrapping_add(self.config.frequency_tuning_word[sign as usize]);
 
-        let amplitude: i16 = match self.config.signal {
-            Signal::Cosine => (dsp::cossin(phase).0 >> 16) as i16,
+        let scale = match self.config.signal {
+            Signal::Cosine => (dsp::cossin(self.phase_accumulator).0 >> 16),
             Signal::Square => {
-                if phase.is_negative() {
-                    i16::MAX
+                if sign {
+                    -i16::MAX as i32
```

Comment:
Stick to what I had in godbolt. It won't overflow due correct bounds on `amplitude`.